### ISSUE TITLE
Build moonlight-chrome as a NW.js app

### ISSUE THREAD

**Issue by BynariStar** [moonlight-stream/moonlight-chrome#355]
With Chrome's impending depreciation of Chrome Apps for all platforms aside from Chrome OS, it appears that soon the only feasible solution to running Moonlight on a PC would be to go back to the old legacy (read: Java) Moonlight app. However, a simple solution to this problem would be to build the current Chrome App as a desktop app, enter NW.js.

NW.js, unlike Electron, actually supports Chrome Apps! Not only that, it supports NaCl too. However, also unlike Electron, it achieves this by essentially being another instance of Chromium. I do think the benefits outweigh the drawbacks of this, however. It's a relatively simple drag-and-drop solution (as much as those are possible in the software development world) to running the current app on the desktop, and it'll do it just as well as Chrome can.

I thought about doing this myself, but I'm not entirely sure how MPEG licensing works when it comes to distribution.

Would be happy to hear what you think of NW.js as a solution, or if you have any alternatives in mind.

**Comment by gruntaway** [comment]
I second this. Would Nw.Js work just as well as the Chrome App? Like, would it be easy to port it and continue the work on this project, or would we have to discontinue the Chrome version altogether?

**Comment by BynariStar** [comment]
As far as I understand, it works exactly the same. NW.js just works as the "Chrome" (pun semi-intended) for the app, so further development of this app would work on both current Chrome and NW.js.

**Comment by apoklyps3** [comment]
would NW.js version bring any improvement of the the resources use?
The problem with the PC version being an exclusive chrome plugin is that it runs with...Chrome, wich is a power hungry browser.
I find myself with over 2GB of ram used on the host machine with only 5 tabs opened. 
No wonder it stuters and freezes all the time when I play on my client (5200U laptop) wich has lower specs...

**Comment by BynariStar** [comment]
I don't have any numbers on resource usage, but my guess is that NW.js does without some of the bloat Chrome has.  I'll try to get some benchmarks on Moonlight running on a clean Chrome instance and on NW.js.

**Comment by BynariStar** [comment]
After some rudimentary tests, I have found that RAM usage in NW.js varies generally between 80 and 400 MB and in Chrome between 100 and 600 MB depending on settings (resolution, bandwidth, etc). Keep in mind these numbers are by no means conclusive, however, they do show that there is still a difference between just running it in Chrome compared to an NW.js app.

**Comment by Jorys-Paulin** [comment]
The most reasonable path for me would be to move this project over to Electron, or even better to WebAssembly.

**Comment by BynariStar** [comment]
Moving to Electron would require a considerable rewrite (or outright replacement) of the [C codebase](https://github.com/moonlight-stream/moonlight-common-c) as mentioned in the initial comment on this issue. In regards to WebAssembly, aside from that too requiring quite a substantial rewrite, it doesn't even seem possible at the moment as mentioned in #251, unless there has been a recent development regarding the missing API.

**Comment by Jorys-Paulin** [comment]
@BynariStar So, should we close this issue?

**Comment by BynariStar** [comment]
@Jorys-Paulin Why? I saw no points against making an NW.js app. You recommended Electron or WASM, but I mentioned why I don't think those are viable options.

**Comment by Jorys-Paulin** [comment]
@BynariStar Well, NW.js is deprecated for me, and the devs mentioned on [this issue](https://github.com/moonlight-stream/moonlight-chrome/issues/371) that they'll probably move on to WebAssembly.

**Comment by BynariStar** [comment]
@Jorys-Paulin What do you mean by deprecated for you?

Also @raidancampbell @cgutman I would be happy to hear your input on this.

**Comment by Jorys-Paulin** [comment]
@BynariStar Well for me Electron allows for more than what NW.js can. I thought I saw I was going deprecated but apparently not.

**Comment by BynariStar** [comment]
@Jorys-Paulin Electron is nice, but it doesn't allow for just packaging the already existing codebase as its own app like NW.js does. The only reason I recommended NW.js was because of the absolute minimal amount of code change necessary to continue having a moonlight app available on desktop platforms.

**Comment by raidancampbell** [comment]
Thanks for the suggestion, I've implemented this on the [feature/native](https://github.com/moonlight-stream/moonlight-chrome/tree/feature/native) branch.

I have very mixed feelings about the whole thing: nwjs is effectively bundling all of chrome at runtime.  I feel dirty using it.  However, it does seem to work: give it a try, I won't prebuild .exe or .app binaries at this stage.

**Comment by jsnar** [comment]
@Jorys-Paulin Electron would be a bad choice because it cannot keep updated with Chromium stable releases as NW.js does. So Electron has lots of security issues unfixed. See the long list of fix security issues  in each release notes of Chrome.

**Comment by Jorys-Paulin** [comment]
@jsnar I would still prefer a WebAssembly version for the future, as discussed in #371, maybe only for the fact that a single version would run on all browsers and platforms.

**Comment by Jorys-Paulin** [comment]
@jsnar But if NW.js follows chrome versions, maybe NaCl modules will also be removed then?

**Comment by gruntaway** [comment]
I completely agree that a WebAssemply version would be preferred, but it might take a while before all the necessary components and APIs are available.

In the meantime NWJS would allow the continuation of the Moonlight Chrome project, which has come a long way and is - by far - the best cross-platform solution we have had to this date.

The only issue I have with it, is that NW.Js normally works by having to "drag&drop" the Chrome App onto the NW.exe file. And that's just not that intuitive, for instance if I wanted to help a friend into Moonlight etc. and he would maybe not understand how to install and run it.

So my question is; Would there possible be a way to "pack" the NW.JS solution somehow, so all you'd have to do to install and launch, would be to open a program (for example, moonlight.exe) and it would automatically run the Chrome App in the NW.JS environment?

**Comment by BynariStar** [comment]
@gruntaway Yes, you can find details on how to build a packaged NW.js app at the [native](https://github.com/moonlight-stream/moonlight-chrome/tree/feature/native) branch mentioned above.

**Comment by md678685** [comment]
Just a thought: would it be worth mentioning in the README or release notes that it's possible to use a prebuilt NW.js SDK build with the packaged extension? I feel that quite a few people would rather not launch Chrome itself in order to use Moonlight, and at the moment it's not entirely obvious to users that you need an SDK build of NW.js to use NaCl.

Assuming npm is installed, the following steps work to install NW.js and run moonlight-chrome (tested on Windows 10):
```
npm install --global nwjs
nw i 0.28.1-sdk
nw moonlight-chrome.crx
```

**Comment by gruntaway** [comment]
Wow @md678685 ! This helped a ton!

Now the last step would be to somehow package it for Windows (and/or other platforms), so it's easy for friends etc. to install and use (without having to bother with explaining what nwjs is and why the moonlight icon doesn't show) :P

**Comment by Jorys-Paulin** [comment]
@gruntaway Actually it's pretty simple, as described in [the readme](https://github.com/moonlight-stream/moonlight-chrome/tree/feature/native#building) you just have to type in `npm run dist`

**Comment by md678685** [comment]
@gruntaway I'd assume that when NW.js is officially supported by Moonlight that it would be packaged up with the build process in the feature/native branch, and so would have the Moonlight icon and would be a single executable. The steps I posted above were intended to serve as a temporary guide for people who can't install the SDKs until there are prebuilt packages available for the native branch.

**Comment by gruntaway** [comment]
Awesome - Thanks guys. :)

For some reason though (perhaps it's just because I don't fully understand how the "native build" works in relation to the normal Chrome project), I get the following error after following the steps:

`  CXX  pnacl/Release/viddec.o
viddec.cpp:488:5: error: cannot initialize a member subobject of type 'DecoderRendererSetup' (aka 'int (*)(int, int,
      int, int, void *, int)') with an lvalue of type 'void (int, int, int, int, void *, int)': different return type
      ('int' vs 'void')
    MoonlightInstance::VidDecSetup,
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [pnacl/Release/viddec.o] Error 1
`

**Comment by gruntaway** [comment]
Nevermind, I figured it out - It was because I had already run "make" (probably months ago), and so all I had to do was run the "npm dist" command.

Now, one funny thing that happens though, is that after running the "npm dist" and everything seems to be built correctly, opening the app (on Mac) results in a weird dialog saying the following: ```
"Your profile can not be used because it is from a newer version of NW.js.

Some features may be unavailable. Please specify a different profile directory or use a newer version of NW.js."
```

Otherwise the App seems to start alright. The icon is missing though, for some reason.

Also a funny quick note: I tried to add a small AJAX call in index.js, just to see if I could ping my locale NAS server at home, but for some reason the NWJS build does not seem to want to execute remote calls. Perhaps some cross-origin policy.

**Comment by raidancampbell** [comment]
@Jorys-Paulin I'd love WebAssembly for many reasons, but the simple lack of raw UDP sockets makes it impossible.  In the meantime, NW.js offers some life support, although I'm not convinced how long they'll retain NaCl support.  We can continue building against the last stable chromium release with NaCl, which is the best solution to my mind right now.

the `feature/native` branch was intended as a proof-of-concept to extend the life beyond the existing Chrome extension.

@gruntaway in regards to the `Your profile can not be used because it is from a newer version of NW.js.` error, I saw this when I used one NW.js version to install (using the "just drop the files in this installed folder method), and a different version that I compiled and ran.  The version mismatch caused state-saving to be incompatible with the two NW.js applications.  You should be able to uninstall one version and the error will be resolved.  If not, there may be some configuration files laying around that need to be deleted.

In regards to the AJAX call: the application is configured to allow UDP [here](https://github.com/moonlight-stream/moonlight-chrome/blob/feature/native/manifest.json#L59), it may be worth trying TCP if you'd like to play around.

**Comment by Jorys-Paulin** [comment]
@raidancampbell How about a middleware that will act on the host as a converter from UDP/TCP sockets to webdockets supported by web standards? How hard would it be?

**Comment by BynariStar** [comment]
I believe this issue is still relevant.

**Comment by Jorys-Paulin** [comment]
Hello! Unfortunately, the chrome version is now deprecated in favour of moonlight-qt.
https://github.com/moonlight-stream/moonlight-qt/releases
